File: 2021/java/120.triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0) return 0;
        int rows = triangle.size();
        int[] dp = new int[rows];
        // init , last row 
        for (int c = 0; c < triangle.get(rows - 1).size(); c++) {
            dp[c] = triangle.get(rows - 1).get(c);
            
        }
        // start from last sencond row 
        for (int r = rows - 2; r >= 0; r--) {
            for (int c = 0; c <= r; c++) {
                dp[c] = Math.min(dp[c], dp[c + 1]) + triangle.get(r).get(c);
            }
        }
        return dp[0];
    }
}

File: 2021/python/120.triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if not triangle or len(triangle) == 0:
            return 0
        below_row = triangle[-1]
        rows = len(triangle)
        for r in reversed(range(rows - 1)):
            for c in range(0, r + 1):
                below_row[c] = min(below_row[c], below_row[c + 1]) + triangle[r][c]
        return below_row[0]
